redundant-bmc/src/redundancy_mgr.cpp

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#include "redundancy_mgr.hpp"
+
+#include <persistent_data.hpp>
+#include <phosphor-logging/lg2.hpp>
+#include <xyz/openbmc_project/Common/error.hpp>
+
+namespace rbmc
+{
+
+RedundancyMgr::RedundancyMgr(sdbusplus::async::context& ctx, Services& services,
+                             Sibling& sibling, RedundancyInterface& iface) :
+    ctx(ctx), services(services), sibling(sibling), redundancyInterface(iface),
+    manualDisable(iface.disable_redundancy_override())
+{
+    try
+    {
+        data::remove(data::key::noRedDetails);
+    }
+    catch (const std::exception& e)
+    {
+        lg2::error("Failed removing NoRedundancyDetails: {ERROR}", "ERROR", e);
+    }
+}
+
+// NOLINTNEXTLINE
+void RedundancyMgr::determineAndSetRedundancy()
+{
+    enableOrDisableRedundancy(getNoRedundancyReasons());
+    redundancyDetermined = true;
+}
+
+redundancy::NoRedundancyReasons RedundancyMgr::getNoRedundancyReasons()
+{
+    redundancy::Input input{
+        .role = redundancyInterface.role(),
+        .siblingPresent = sibling.isBMCPresent(),
+        .siblingHeartbeat = sibling.hasHeartbeat(),
+        .siblingProvisioned = sibling.getProvisioned().value_or(false),
+        .siblingHasSiblingComm = sibling.getSiblingCommsOK().value_or(false),
+        .siblingRole = sibling.getRole().value_or(Role::Unknown),
+        .siblingState = sibling.getBMCState().value_or(BMCState::NotReady),
+        .codeVersionsMatch =
+            services.getFWVersion() == sibling.getFWVersion().value_or(""),
+        .manualDisable = manualDisable};
+
+    auto reasons = redundancy::getNoRedundancyReasons(input);
+
+    std::map<redundancy::NoRedundancyReason, std::string> details;
+
+    std::ranges::transform(
+        reasons, std::inserter(details, details.begin()),
+        [](const auto& reason) {
+            auto desc = redundancy::getNoRedundancyDescription(reason);
+            lg2::info("No redundancy because: {DESC}", "DESC", desc);
+            return std::pair{reason, desc};
+        });
+
+    try
+    {
+        data::write(data::key::noRedDetails, details);
+    }
+    catch (const std::exception& e)
+    {
+        lg2::error("Failed serializing NoRedundancyReasons: {ERROR}", "ERROR",
+                   e);
+    }
+
+    return reasons;
+}
+
+void RedundancyMgr::enableOrDisableRedundancy(
+    const redundancy::NoRedundancyReasons& disableReasons)
+{
+    auto enable = disableReasons.empty();
+
+    if (enable)
+    {
+        lg2::info("Enabling redundancy");
+    }
+    else
+    {
+        lg2::info("Redundancy must be disabled");
+    }
+
+    redundancyInterface.redundancy_enabled(enable);
+}
+
+void RedundancyMgr::disableRedPropChanged(bool disable)
+{
+    if (services.isPoweredOn())
```

Comment:
try..catch is missing? https://github.com/ibm-openbmc/phosphor-state-manager/pull/73/commits/91a45daf3f4a99c222dabe22364b8cdc8982ecee#diff-d3dae29cf343c55b54ccd48d1feb447d1950d3d0a836d135b9492dca5d843d94

Reply:
Good Catch.